### PR TITLE
[core] fix thread safe issue in #19832

### PR DESCRIPTION
**Pull request by tiger100256-hu**
### Details:
 - *I lost code `t_stream_count_map[(void*)this] = item.first;` in #19832*
 - *the thread safe issue happen in below workflow*
 - create thread A
   call CustomThreadLocal:local() in thread A -> create stream A (the count of stream A is 2) 
   destory thread A (the count of stream A is 1)
   create thread B (same thread id with thread A)
   call CustomThreadLocal:local() in thread B -> use stream A(the count of stream A is 1, so it's broken)
  - *add testcase, also fix  https://github.com/openvinotoolkit/openvino/pull/19986/files#r1332774754*

### Tickets:
 - Closes https://github.com/openvinotoolkit/openvino/issues/24509

### CODE REVIEW

**Comment by riverlijunjie** [comment]
Can we add test cases for 2 models to concurrently create infer and do infer, including sync infer and async infer?

**Comment by tiger100256-hu** [comment]
@riverlijunjie  ok, I will update the test case

**Comment by tiger100256-hu** [comment]
@riverlijunjie  already add test case for two model and async infer, please help to continue review

File: tests/stress_tests/common/ie_pipelines/pipelines.cpp [hunk 1]
```diff
@@ -99,15 +99,13 @@ std::function<void()> reinfer_request_inference(std::shared_ptr<InferApiBase> &i
...
-            ie_wrapper->prepare_input();
-            ie_wrapper->infer();
+std::function<void()> recreate_and_infer_in_thread(std::shared_ptr<InferApiBase> &ie_wrapper, const bool async) {
+    return [async, &ie_wrapper] {
+        auto func = [&ie_wrapper, &async] {
+            ie_wrapper->create_and_infer(async);
         };
         std::thread t(func);
-	t.join();
+	    t.join();
```
Comment (riverlijunjie) [inline tests/stress_tests/common/ie_pipelines/pipelines.cpp:108]: Remove whitespace here.

**Review by riverlijunjie — APPROVED** [review]
LGTM

File: tests/stress_tests/common/ie_pipelines/pipelines.cpp (hunk 1 above)
Comment (tiger100256-hu) [reply tests/stress_tests/common/ie_pipelines/pipelines.cpp:108]: ok, I will update it

**Review by sunxiaoxia2022 — APPROVED** [review]

File: tests/stress_tests/common/ie_pipelines/pipelines.cpp (hunk 1 above)
Comment (tiger100256-hu) [reply tests/stress_tests/common/ie_pipelines/pipelines.cpp]: updated

**Review by wangleis — APPROVED** [review]

**Pull request merged**